### PR TITLE
ipn: apply tailnet-wide default for auto-updates

### PR DESCRIPTION
**Pull request by awly**
When auto-update setting in local Prefs is unset, apply the tailnet
default value from control. This only happens once, when we apply the
default (or when the user manually overrides it), tailnet default no
longer affects the node.

Updates #16244

### CODE REVIEW

File: ipn/prefs.go [hunk 1]
```diff
@@ -237,7 +238,17 @@ type AutoUpdatePrefs struct {
 	// Apply specifies whether background auto-updates are enabled. When
 	// enabled, tailscaled will apply available updates in the background.
 	// Check must also be set when Apply is set.
-	Apply bool
+	Apply opt.Bool
+}
+
+func (au1 AutoUpdatePrefs) Equals(au2 AutoUpdatePrefs) bool {
+	// This could almost be as easy as `au1.Apply == au2.Apply`, except that
+	// opt.Bool("") and opt.Bool("unset") should be treated es equal.
```
Comment (noncombatant) [inline ipn/prefs.go:246]: typo: "as equal"

File: types/opt/bool.go [hunk 2]
```diff
@@ -18,6 +18,16 @@ import (
 // field without it being dropped.
 type Bool string
 
+// NewBool constructs a new Bool value equal to b. The returned Bool is set,
+// unless Set("") or Clear() methods are called.
+func NewBool(b bool) Bool {
+	if b {
```
Comment (noncombatant) [inline types/opt/bool.go:24]: Can this function be simplified to look more like `Set`, below? I.e. use `strconv`?

**Review by noncombatant — APPROVED** [review]

File: types/opt/bool.go (hunk 2 above)
Comment (awly) [reply types/opt/bool.go]: Duh, yeah that's cleaner!

File: ipn/ipnlocal/local.go [hunk 3]
```diff
@@ -2495,6 +2497,31 @@ func (b *LocalBackend) onClientVersion(v *tailcfg.ClientVersion) {
...
+	if prefs == nil {
+		b.logf("[unexpected]: received tailnet default auto-update callback but current prefs are nil")
+		return
+	}
+	if _, ok := prefs.AutoUpdate.Apply.Get(); ok {
+		// Apply was already set from a previous default or manually by the
+		// user. Tailnet default should not affect us, even if it changes.
+		return
+	}
+	b.logf("using tailnet default auto-udpate setting: %v", au)
```
Comment (maisem) [inline ipn/ipnlocal/local.go:2511]: auto-update

File: tailcfg/tailcfg.go [hunk 4]
```diff
@@ -1877,6 +1877,11 @@ type MapResponse struct {
 	// download and whether the client is using it. A nil value means no change
 	// or nothing to report.
 	ClientVersion *ClientVersion `json:",omitempty"`
+
+	// DefaultAutoUpdate is the default node auto-update setting for this
+	// tailnet. The node is free to opt-in or out locally regardless of this
+	// value.
+	DefaultAutoUpdate opt.Bool `json:",omitempty"`
```
Comment (maisem) [inline tailcfg/tailcfg.go:1884]: bump CapVer

File: tailcfg/tailcfg.go [hunk 5]
```diff
@@ -1877,6 +1877,11 @@ type MapResponse struct {
 	// download and whether the client is using it. A nil value means no change
 	// or nothing to report.
 	ClientVersion *ClientVersion `json:",omitempty"`
+
+	// DefaultAutoUpdate is the default node auto-update setting for this
+	// tailnet. The node is free to opt-in or out locally regardless of this
```
Comment (maisem) [inline tailcfg/tailcfg.go:1882]: clarify when it takes affect

File: ipn/ipnlocal/local.go [hunk 6]
```diff
@@ -2495,6 +2497,31 @@ func (b *LocalBackend) onClientVersion(v *tailcfg.ClientVersion) {
 	b.send(ipn.Notify{ClientVersion: v})
 }
 
+func (b *LocalBackend) onTailnetDefaultAutoUpdate(au bool) {
+	prefs := b.pm.CurrentPrefs().AsStruct()
```
Comment (maisem) [inline ipn/ipnlocal/local.go:2501]: you can defer the clone here till after line 2509

File: ipn/ipnlocal/local_test.go [hunk 7]
```diff
@@ -2055,3 +2056,46 @@ func TestPreferencePolicyInfo(t *testing.T) {
...
 }
+
+func TestOnTailnetDefaultAutoUpdate(t *testing.T) {
+	tests := []struct {
+		desc           string
+		before, after  opt.Bool
+		tailnetDefault bool
+	}{
+		{
+			before:         opt.Bool("unset"),
```
Comment (maisem) [inline ipn/ipnlocal/local_test.go:2067]: might want to test `opt.Bool("")` too

File: ipn/prefs.go [hunk 8]
```diff
@@ -600,7 +611,7 @@ func NewPrefs() *Prefs {
 		NetfilterMode:    preftype.NetfilterOn,
 		AutoUpdate: AutoUpdatePrefs{
 			Check: true,
-			Apply: false,
+			Apply: opt.Bool("unset"),
```
Comment (maisem) [inline ipn/prefs.go:614]: ah

**Pull request merged**